### PR TITLE
backup: add backup endpoint

### PR DESCRIPTION
**Pull request by overvenus**
###  What have you changed?

Add backup endpoint, the endpoint coordinates backup tasks and dispatches them to different workers.

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

Unit tests.

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No.

### CODE REVIEW

File: src/storage/kv/rocksdb_engine.rs [hunk 1]
```diff
@@ -260,11 +265,18 @@ impl Engine for RocksEngine {
...
         )));
-        fail_point!("rockskv_async_snapshot_not_leader", |_| {
-            let mut header = ErrorHeader::default();
+        let not_leader = {
+            let mut header = kvproto::errorpb::Error::new();
             header.mut_not_leader().set_region_id(100);
-            Err(Error::Request(header))
+            header
+        };
+        let _not_leader = not_leader.clone();
```
Comment (5kbpers) [inline src/storage/kv/rocksdb_engine.rs]: Is it necessary to declare a new variable?

**Comment by 5kbpers** [comment]
reset LGTM

File: src/storage/kv/rocksdb_engine.rs (hunk 1 above)
Comment (overvenus) [reply src/storage/kv/rocksdb_engine.rs]: Yes, the `failpoints` macro moves `not_leader`.

File: components/backup/src/endpoint.rs [hunk 2]
```diff
@@ -74,29 +96,317 @@ impl Task {
...
+            .clone()
+            .map_or_else(Vec::new, |k| k.into_encoded());
+        let res = self.region_info.seek_region(
+            &start_key_,
+            Box::new(move |iter| {
+                for info in iter {
+                    let region = &info.region;
+                    if end_key.is_some() {
+                        let end_slice = end_key.as_ref().unwrap().as_encoded().as_slice();
+                        if end_slice <= region.get_start_key() {
```
[the remark this answers is not in the document]
Comment (overvenus) [reply components/backup/src/endpoint.rs]: Well, I think to flatten the logical is more clear than a function.

**Comment by overvenus** [comment]
PTAL, thanks!

**Pull request merged**